Product/Production/Services/AuditRepositoryCore/src/main/java/gov/hhs/fha/nhinc/audit/transform/AuditTransforms.java

```diff
@@ -251,25 +253,72 @@ private EventIdentificationType getEventIdentificationType(CodedValueType eventI
         return eventIdentification;
     }
 
-    private AuditMessageType.ActiveParticipant getActiveParticipantSource(boolean isRequesting,
+    /**
+     * This method builds ActiveParticipant Source object
+     *
+     * @param target
+     * @param serviceName
+     * @param isRequesting
+     * @param webContextProperties
+     * @return
+     */
+    private ActiveParticipant getActiveParticipantSource(NhinTargetSystemType target, String serviceName, boolean isRequesting,
         Properties webContextProperties) {
+        ActiveParticipant participant = new ActiveParticipant();
+        String hostAddress = null;
+
+        /* if Retrieve Document Service, Activeparticipant Source and destination is same on both initiator and responder side. And for this service, source is considered who generates the document, meaning that destination that generates the document is considered source for RD and vice versa for destination details for ActiveParticipant */
+        if (serviceName.equals(NhincConstants.DOC_RETRIEVE_SERVICE_NAME)) {
+            hostAddress = getWebServiceUrlFromRemoteObject(target, serviceName);
+            if (hostAddress != null) {
+                try {
+                    URL url = new URL(hostAddress);
+                    participant.setUserID(hostAddress);
+                    hostAddress = url.getHost();
+                    participant.setNetworkAccessPointID(hostAddress);
+                    participant.setNetworkAccessPointTypeCode(getNetworkAccessPointTypeCode(hostAddress));
+                } catch (MalformedURLException ex) {
+                    LOG.error(ex);
+                    // The url is null or not a valid url; for now, set the user id to anonymous
+                    participant.setUserID(AuditTransformsConstants.ACTIVE_PARTICIPANT_USER_ID_SOURCE);
+                    // TODO: For now, hardcode the value to localhost; need to find out if this needs to be set
```

Comment:
Has this todo been handled already?  Or do we need to open another ticket for this?  We just don't want to lose track.